### ISSUE TITLE
MNIST Tutorial Appears to Not Toggle XLA Compilation

### ISSUE THREAD

**Issue by lwogulis** [tensorflow/tensorflow#9579]
I have been using the JIT compilation/XLA tutorial (https://www.tensorflow.org/performance/xla/jit#step_3_run_with_xla), and it seems that whether or not XLA compilation happens doesn't depend on the statement on line 63 in mnist_softmax_xla.py. The comment above it says that line will turn on XLA JIT compilation. When I run with the two options explained on the page (--xla='' for compiling without XLA, and TF_XLA_FLAGS=--xla_generate_hlo_graph=.* for compiling with XLA), the second executes line 63 and the first does not. But, both seem to compute in exactly the same way, going through compiler/xla/service. The output for both runs is:
```
Extracting /tmp/tensorflow/mnist/input_data/train-images-idx3-ubyte.gz
Extracting /tmp/tensorflow/mnist/input_data/train-labels-idx1-ubyte.gz
Extracting /tmp/tensorflow/mnist/input_data/t10k-images-idx3-ubyte.gz
Extracting /tmp/tensorflow/mnist/input_data/t10k-labels-idx1-ubyte.gz
2017-05-01 12:50:25.203870: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use AVX2 instructions, but these are available on your machine and could speed up CPU computations.
2017-05-01 12:50:25.203904: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use FMA instructions, but these are available on your machine and could speed up CPU computations.
2017-05-01 12:50:25.232661: I tensorflow/compiler/xla/service/platform_util.cc:58] platform Host present with 8 visible devices
2017-05-01 12:50:25.232702: I tensorflow/compiler/xla/service/platform_util.cc:58] platform Xpu present with 8 visible devices
2017-05-01 12:50:25.233155: I tensorflow/compiler/xla/service/service.cc:180] XLA service executing computations on platform Host. Devices:
2017-05-01 12:50:25.233165: I tensorflow/compiler/xla/service/service.cc:187]   StreamExecutor device (0): <undefined>, <undefined>
2017-05-01 12:50:25.233894: I tensorflow/compiler/xla/service/platform_util.cc:58] platform Host present with 8 visible devices
2017-05-01 12:50:25.233903: I tensorflow/compiler/xla/service/platform_util.cc:58] platform Xpu present with 8 visible devices
2017-05-01 12:50:25.234360: I tensorflow/compiler/xla/service/service.cc:180] XLA service executing computations on platform Xpu. Devices:
2017-05-01 12:50:25.234369: I tensorflow/compiler/xla/service/service.cc:187]   StreamExecutor device (0): <undefined>, <undefined>
2017-05-01 12:50:25.234372: I tensorflow/compiler/xla/service/service.cc:187]   StreamExecutor device (1): <undefined>, <undefined>
2017-05-01 12:50:25.234376: I tensorflow/compiler/xla/service/service.cc:187]   StreamExecutor device (2): <undefined>, <undefined>
2017-05-01 12:50:25.234379: I tensorflow/compiler/xla/service/service.cc:187]   StreamExecutor device (3): <undefined>, <undefined>
2017-05-01 12:50:25.234382: I tensorflow/compiler/xla/service/service.cc:187]   StreamExecutor device (4): <undefined>, <undefined>
2017-05-01 12:50:25.234385: I tensorflow/compiler/xla/service/service.cc:187]   StreamExecutor device (5): <undefined>, <undefined>
2017-05-01 12:50:25.234389: I tensorflow/compiler/xla/service/service.cc:187]   StreamExecutor device (6): <undefined>, <undefined>
2017-05-01 12:50:25.234392: I tensorflow/compiler/xla/service/service.cc:187]   StreamExecutor device (7): <undefined>, <undefined>
0.9206
```

**Comment by hawkinsp** [comment]
Those are informational log messages that are emitted during startup. They do not indicate any code is actually running via XLA.

Try setting the environment variable TF_CPP_MIN_VLOG_LEVEL=2 and looking for log messages from xla_device_launch_op.cc or xla_local_launch_op.cc.

**Comment by lwogulis** [comment]
Okay, when I do that, there is quite a lot printed, but neither output has anything from either of those files. I'm running:
`TF_XLA_FLAGS=--xla_generate_hlo_graph=.* python mnist.py &> out_xla`
and
`python mnist.py --xla='' &> out_xla`
and searching for 'xla_device_launch_op' and 'xla_local_launch_op' in the file out_xla each time, and neither is found.

**Comment by drpngx** [comment]
Just to double-check: did you configure with XLA?

**Comment by lwogulis** [comment]
Yes, I set TF_ENABLE_XLA=1 when I configure

**Comment by hawkinsp** [comment]
I think this is working as intended. By default, only the GPU JIT will be enabled by the --xla flag to that script. We do not enable the CPU JIT by default at the moment because it does not support inter-operator parallelism. And you do not have a GPU installed or are not compiling with CUDA.

If you wish to enable XLA on the CPU, you will need to modify the mnist_softmax_xla.py script to use a `with tf.device("device:XLA_CPU:0"):` scope to force CPU operators to be compiled with XLA. The "Placing operators on XLA devices" section of the tutorial describes how to do this.

(I cannot speak to what is happening with your custom "Xpu" device --- if I cannot see the source code, there is no way for me to debug it.)

**Comment by lwogulis** [comment]
I am back to trying to run the mnist example with XLA, and I just tried your suggestion of forcing "with tf.device('device:XLA_CPU:0')". I get the error that it could not satisfy the explicit device specification. Am I putting the statement in the wrong place?
```
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

import argparse
import sys 

import tensorflow as tf

from tensorflow.examples.tutorials.mnist import input_data
from tensorflow.python.client import timeline

FLAGS = None


def main(_):
  # Import data
  mnist = input_data.read_data_sets(FLAGS.data_dir, one_hot=True)

  # Create the model
  x = tf.placeholder(tf.float32, [None, 784])
  w = tf.Variable(tf.zeros([784, 10]))
  b = tf.Variable(tf.zeros([10]))
  y = tf.matmul(x, w) + b 

  # Define loss and optimizer
  y_ = tf.placeholder(tf.float32, [None, 10])

  # The raw formulation of cross-entropy,
  #
  #   tf.reduce_mean(-tf.reduce_sum(y_ * tf.log(tf.nn.softmax(y)),
  #                                 reduction_indices=[1]))
  #
  # can be numerically unstable.
  #
  # So here we use tf.nn.softmax_cross_entropy_with_logits on the raw
  # outputs of 'y', and then average across the batch.
  cross_entropy = tf.reduce_mean(
      tf.nn.softmax_cross_entropy_with_logits(labels=y_, logits=y))
  train_step = tf.train.GradientDescentOptimizer(0.5).minimize(cross_entropy)

  config = tf.ConfigProto()
  jit_level = 0 
  if FLAGS.xla:
    # Turns on XLA JIT compilation.
    jit_level = tf.OptimizerOptions.ON_1
    print('XLA on')
  config.graph_options.optimizer_options.global_jit_level = jit_level
  run_metadata = tf.RunMetadata()
  with tf.device('device:XLA_CPU:0'):
      sess = tf.Session(config=config)
      tf.global_variables_initializer().run(session=sess)
      # Train
      train_loops = 1000
      for i in range(train_loops):
        batch_xs, batch_ys = mnist.train.next_batch(100)

        # Create a timeline for the last loop and export to json to view with
        # chrome://tracing/.
        if i == train_loops - 1:
          sess.run(train_step,
                   feed_dict={x: batch_xs,
                              y_: batch_ys},
                   options=tf.RunOptions(trace_level=tf.RunOptions.FULL_TRACE),
                   run_metadata=run_metadata)
          trace = timeline.Timeline(step_stats=run_metadata.step_stats)
          with open('timeline.ctf.json', 'w') as trace_file:
            trace_file.write(trace.generate_chrome_trace_format())
        else:
          sess.run(train_step, feed_dict={x: batch_xs, y_: batch_ys})

      # Test trained model
      correct_prediction = tf.equal(tf.argmax(y, 1), tf.argmax(y_, 1))
      accuracy = tf.reduce_mean(tf.cast(correct_prediction, tf.float32))
      print(sess.run(accuracy,
                     feed_dict={x: mnist.test.images,
                                y_: mnist.test.labels}))
      sess.close()


if __name__ == '__main__':
  parser = argparse.ArgumentParser()
  parser.add_argument(
      '--data_dir',
      type=str,
      default='/tmp/tensorflow/mnist/input_data',
      help='Directory for storing input data')
  parser.add_argument(
      '--xla', type=bool, default=True, help='Turn xla via JIT on')
  FLAGS, unparsed = parser.parse_known_args()
  tf.app.run(main=main, argv=[sys.argv[0]] + unparsed)
```

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by MarkSonn** [comment]
Did you manage to solve the issue? I'm also having this problem (running Ubuntu 16.04 with 2x 1080 Ti GPUs)

**Comment by MarkSonn** [comment]
@aselle  @hawkinsp I have followed the above instructions, what's next? Any help would be really appreciated 👍

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @hawkinsp: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @hawkinsp: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @hawkinsp: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @hawkinsp: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @hawkinsp: It has been 145 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @hawkinsp: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @hawkinsp: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @hawkinsp: It has been 45 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by vishalsubbiah** [comment]
So I found replacing tf.Variable with tf.get_variable and setting use_resource to True works. Also the flag command is TF_XLA_FLAGS="--xla_hlo_graph_path=PATH --xla_generate_hlo_graph=.*"

**Comment by hawkinsp** [comment]
I'm not sure this issue is still relevant. Feel free to reopen if it still occures.
